src/libraries/System.Private.Uri/src/System/IriHelper.cs

```diff
@@ -69,62 +69,55 @@ internal static bool CheckIriUnicodeRange(uint value, bool isQuery)
         public static bool IsInInclusiveRange(uint value, uint min, uint max)
             => (value - min) <= (max - min);
 
-        //
         // IRI normalization for strings containing characters that are not allowed or
         // escaped characters that should be unescaped in the context of the specified Uri component.
-        //
-        internal static unsafe string EscapeUnescapeIri(char* pInput, int start, int end, bool isQuery)
+        internal static string EscapeUnescapeIri(ReadOnlySpan<char> span, bool isQuery)
         {
-            Debug.Assert(end >= 0 && start >= 0 && start <= end);
-
-            int size = end - start;
-            var dest = size <= Uri.StackallocThreshold
+            var dest = span.Length <= Uri.StackallocThreshold
                 ? new ValueStringBuilder(stackalloc char[Uri.StackallocThreshold])
-                : new ValueStringBuilder(size);
+                : new ValueStringBuilder(span.Length);
 
             Span<byte> maxUtf8EncodedSpan = stackalloc byte[4];
 
-            for (int i = start; i < end; ++i)
+            for (int i = 0; (uint)i < (uint)span.Length; i++)
             {
-                char ch = pInput[i];
+                char ch = span[i];
+
                 if (ch == '%')
                 {
-                    if (end - i > 2)
+                    if ((uint)(i + 2) < (uint)span.Length)
```

Comment:
```suggestion
                    if ((uint)i < (uint)(span.Length - 2))
```
To avoid re-computing `i + 2` on every iteration. But does this also elide the bound check?

Reply:
@EgorBo can the JIT learn to elide the bound checks here (or in [M2 in this demo](https://godbolt.org/z/3zhnbc761))?